### PR TITLE
ARROW-1453: [C++/Python] Support non-contiguous Tensors in WriteTensor

### PR DESCRIPTION
**Pull request by pcmoritz**

### CODE REVIEW

**Comment by robertnishihara** [comment]
Looks good to me.
- either python linting isn't being applied to python/pyarrow/tests/test_serialization.py or the linting is just much less strict (e.g., things like `data[9,0,0] = 1.0` should probably have more spaces, also the functions should have two newlines before and after)
- the optimization I would consider doing (in a later PR) is to figure out the size of the contiguous blocks that can be written all at once with a `memcpy`. If that size is greater than a certain quantity, then use the recursive approach here but terminating with a bulk `memcpy`. If it is less than a certain quantity (for example if it is just 1) then maybe abandon the recursive approach and just iterate over the elements without recursion.

**Comment by pcmoritz** [comment]
+1 for merging

I fixed the linting and created a JIRA for the optimization here: https://issues.apache.org/jira/browse/ARROW-1457 There is also a comment about it in the code.

The Ray test suite now passes on top of this commit, so once this is merged we should be ready for 0.7 from our side: https://github.com/ray-project/ray/pull/910

**Comment by wesm** [comment]
There's an issue where linting is not failing the build: https://issues.apache.org/jira/browse/ARROW-1430

File: cpp/src/arrow/ipc/writer.cc [hunk 1]
```diff
@@ -563,23 +563,48 @@ Status WriteLargeRecordBatch(const RecordBatch& batch, int64_t buffer_start_offs
                           pool, kMaxNestingDepth, true);
 }
 
-Status WriteTensor(const Tensor& tensor, io::OutputStream* dst, int32_t* metadata_length,
-                   int64_t* body_length) {
-  if (!tensor.is_contiguous()) {
-    return Status::Invalid("No support yet for writing non-contiguous tensors");
+Status WriteStridedTensorData(int dim_index, int64_t offset, int elem_size,
+                              const Tensor& tensor, io::OutputStream* dst) {
+  // TODO(pcm): optimize recursion base case by doing bulk write if possible
```
Comment (wesm) [inline cpp/src/arrow/ipc/writer.cc]: Let me go ahead and do this. We can allocate a single scratch space buffer for use for each of the inner writes

**Comment by wesm** [comment]
Appveyor build running here: https://ci.appveyor.com/project/pcmoritz/arrow/build/1.0.424. The integration test entry in the build matrix has been flaky because of issues with the LLVM apt repo, so we can ignore that failure

+1, will merge once the builds look good. Feel free to resolve ARROW-1457 if my fix seems adequate

**Pull request closed without merging**